Support set auto-self-update mode (#2763)
rust-lang/rustup: doc/src/basics.md

## Keeping `rustup` up to date

## Keeping rustup up to date

It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
variable.  This setting supports three values: `enable` and `disable` and `check-only`.

* `disable` will ensure that no automatic self updating actions are taken.
* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.

> Whether automatic self-update is enabled or not, you can request that Rustup check for, and update it self to the

Reviewer: ```suggestion
Whether automatic self-update is enabled or not, you can request that Rustup check for, and update itself to the
```

Author: Fixed.

latest version of `rustup` without updating any installed toolchains by running `rustup
self update`:

```console
$ rustup self update
info: checking for self-updates
info: downloading self-updates
```

**Note**: If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
toolchain installation as well.  You can prevent this automatic behaviour by
passing the `--no-self-update` argument when running `rustup update` or

> `rustup toolchain install`.

Reviewer: A previous change deliberately called this out in a block quote with its own low-tier heading.  I suggest you don't undo that since it's a good readability improvement.

Author: Added.

## Keeping `rustup` up to date

## Keeping rustup up to date

It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
variable.  This setting supports three values: `enable` and `disable` and `check-only`.

* `disable` will ensure that no automatic self updating actions are taken.
* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.

Whether automatic self-update is enabled or not, you can request that Rustup check for, and update itself to the
latest version of `rustup` without updating any installed toolchains by running `rustup

> self update`:

Reviewer 2: I think this section also needs to mention the feature that can be disable that disables this capability entirely.

Author: It was mentioned at the very beginning.

```console
$ rustup self update
info: checking for self-updates
info: downloading self-updates
```

> ### Disable automatic self-updates with the parameter of the command

> > If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any

Reviewer 2: This should probably mention that the default is `enable`?

Author: Mentioned.

> ### Disable automatic self-updates with the parameter of the command
> If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
> toolchain installation as well.  You can prevent this automatic behaviour by
> passing the `--no-self-update` argument when running `rustup update` or
> `rustup toolchain install`.

Reviewer: You still seem to have removed the blockquote (`>`) here which was added to provide callout clarity.

```suggestion
> ### Disable automatic self-updates with the parameter of the command
> If auto-self-update is `enable`, `rustup` will automatically update itself at the end of any
> toolchain installation as well.  You can prevent this automatic behaviour by
> passing the `--no-self-update` argument when running `rustup update` or
> `rustup toolchain install`.
```

Author: Sorry I missed it. 
Added. Thanks for your review!

## Keeping `rustup` up to date

> If your `rustup` build with the `no-self-update` feature, it will not be able to update itself.

Reviewer 2: Grammar fix: "was built with"

Reviewer 2: Suggest adding that this is the default and that the rustups available from https://rustup.rs are built this way.

> If your `rustup` build without the `no-self-update` feature, 

Reviewer 2: same grammar fix needed here

it is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration variable. 
This setting supports three values: `enable` and `disable` and `check-only`.

* `disable` will ensure that no automatic self updating actions are taken.

> * `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.

Reviewer 2: check for should not be hypenated here. Just say "commands will also check for, and install, any new version of Rustup.

* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.

Whether `auto-self-update` is `enable` or not, you can request that Rustup check for, and update itself to the
latest version of `rustup` without updating any installed toolchains by running `rustup
self update`:

```console
$ rustup self update
info: checking for self-updates
info: downloading self-updates
```

> > ### Disable automatic self-updates with the parameter of the command

Reviewer 2: The conditional clause makes this hard to process. Perhaps we start over now things are different. How about
```
### Disabling self updates on a per-invocation basis
> Self updates can also be suppressed on individual invocations of `rustup` by passing the argurment `--no-self-update`  when running `rustup update` or `rustup toolchain install`.
```

## Keeping `rustup` up to date

> If your `rustup` was build with the `no-self-update` feature, it will not be able to update itself.

Reviewer 2: should be past tense 

'was built with'

## Keeping rustup up to date

> We control the automatic update of rustup by setting the auto-self-update configuration. 
> The auto-self-update setting supports three values: `enable` and `disable` and `check-only`.

Reviewer: ```suggestion
It is possible to control Rustup's automatic self update mechanism with the `auto-self-update` configuration
variable.  This setting supports three values: `enable` and `disable` and `check-only`.
```

> When it is `enable`, running `rustup update` will also check for updates to `rustup` and automatically install the latest version.
> If it is `disable`, the available updates is not checked and installed. 
> In addition, you can set auto-self-update to `check-only` if you want to check rustup for available updates during `rustup update` as well.

Reviewer: ```suggestion
* `disable` will ensure that no automatic self updating actions are taken.
* `enable` will mean that `rustup update` and similar commands will also check-for, and install, any update to Rustup.
* `check-only` will cause any automatic self update to check and report on any updates, but not to automatically install them.
```

> To manually force check for updates and install the

Reviewer: ```suggestion
Whether automatic self-update is enabled or not, you can request that Rustup check for, and update it self to the
```

> latest version of `rustup` without updating installed toolchains type `rustup

Reviewer: ```suggestion
latest version of `rustup` without updating any installed toolchains by running `rustup
```
